File: Searching and Sorting Algorithms/codigo_cs/15MergeSort.cs
using System;
using System.Linq;

public class Test15

{

    static void Merge(int[] arr, int l, int m, int r)
    {
        int i, j;
        int[] left = new int[m - l + 1];
        int[] right = new int[r - m];
        for (i = 0; i < left.Length; i++)
            left[i] = arr[l + i];
        for (j = 0; j < right.Length; j++)
            right[j] = arr[m + 1 + j];
        int k = l;
        i = j = 0;
        while (i < left.Length && j < right.Length)
        {
            if (left[i] <= right[j])
                arr[k++] = left[i++];
            else
                arr[k++] = right[j++];
        }
        while (i < left.Length)
            arr[k++] = left[i++];
        while (j < right.Length)
            arr[k++] = right[j++];


    }

    public static void MergeSort(int[] arr, int l, int r)
    {
        if (l == r)
            return;
        int mid = (l + r) / 2;
        MergeSort(arr, l, mid);
        MergeSort(arr, mid + 1, r);
        Merge(arr, l, mid, r);

    }
}

File: Searching and Sorting Algorithms/15MergeSort.py
def merge(arr,l,m,r):
    l_a=[0]*(m+1-l)
    r_a=[0]*(r-m)
    for i in range(m+1-l):
        l_a[i]=arr[l+i]
    for j in range(r-m):
        r_a[j]=arr[m+1+j]
    i=j=0
    k=l
    while(i<(m+1-l) and j<(r-m)):
        if l_a[i]<r_a[j]:
            arr[k]=l_a[i]
            i+=1
        else:
            arr[k]=r_a[j]
            j+=1
        k+=1
    while i<(m+1-l):
        arr[k]=l_a[i]
        i+=1
        k+=1

    while j<(r-m):
        arr[k]=r_a[j]
        j+=1
        k+=1



def merge_sort(arr,l,r):
    if l==r:
        return
    mid=(l+r)//2
    merge_sort(arr,l,mid)
    merge_sort(arr,mid+1,r)
    merge(arr,l,mid,r)

arr=[6,5,3,1,8,7,2,4]
merge_sort(arr,0,len(arr)-1)
print(arr)